graphql_compiler/compiler/emit_sql.py

```diff
@@ -74,31 +77,249 @@ def _find_columns_used(sql_schema_info, ir):
+
+    def visit_traversed_vertex(self, join_descriptor, from_table, to_table):
+        """Add a new join descriptor for every vertex traversed in the fold."""
+        self._join_info.append((join_descriptor, from_table, to_table))
+
+    def end_fold(self, alias_generator, from_clause, outer_from_table):
+        """Produce the final subquery and join it onto the rest of the query."""
+        # for now we only handle folds containing one traversal (i.e. join)
+        if len(self.join_info) > 1:
+            raise NotImplementedError('Folds containing multiple traversals are not implemented.')
+        edge, from_alias, to_alias = self.join_info.pop()
+
+        # produce the from clause/joins for the subquery
+        subquery_from_clause = self._construct_fold_joins(edge, from_alias, to_alias)
+
+        # produce full subquery
+        fold_subquery = self._construct_fold_subquery(subquery_from_clause).alias(
+            alias_generator.generate_subquery()
+        )
+
+        # join the subquery onto the rest of the query
+        joined_from_clause = sqlalchemy.join(
+            from_clause,
+            fold_subquery,
+            onclause=(outer_from_table.c[edge.from_column] == fold_subquery.c[edge.from_column]),
+            isouter=True
+        )
+
+        return fold_subquery, joined_from_clause, outer_from_table
+
+
+class UniqueAliasGenerator(object):
+    """Mutable class used to generate unique aliases for subqueries."""
+
+    def __init__(self):
+        """Create unique subquery aliases by tracking counter."""
+        self._fold_count = 1
+
+    def generate_subquery(self):
+        """Generate a new subquery alias and increment the counter."""
+        alias = 'folded_subquery_{}'.format(self._fold_count)
+        self._fold_count += 1
+        return alias
+
+
 class CompilationState(object):
     """Mutable class used to keep track of state while emitting a sql query."""
 
     def __init__(self, sql_schema_info, ir):
         """Initialize a CompilationState, setting the current location at the root of the query."""
-        # Metadata
+        # Immutable metadata
         self._sql_schema_info = sql_schema_info
         self._ir = ir
-        self._used_columns = _find_columns_used(sql_schema_info, ir)
+        self._used_columns = _find_columns_used_outside_folds(sql_schema_info, ir)
+        self._all_folded_outputs = _find_folded_outputs(ir)
 
         # Current query location state. Only mutable by calling _relocate.
         self._current_location = None  # the current location in the query. None means global.
         self._current_alias = None  # a sqlalchemy table Alias at the current location
-        self._aliases = {}  # mapping marked query paths to table _Aliases representing them
+
+        # Dict mapping (some_location.query_path, some_location.fold_path) tuples to
+        # corresponding table _Aliases. some_location is either self._current_location
+        # or an open fold scope location.
```

Comment:
i still think this comment is a bit off. a fold scope location doesn't have a `query_path`. `some_location` is the `base_location` of an open fold scope location